File: tests/algorithms/transpiler/FS/maths/solovay_strassen_primality_test.fs
// Generated 2025-08-17 12:28 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let rec _str v =
    match box v with
    | :? float as f -> sprintf "%.10g" f
    | _ ->
        let s = sprintf "%A" v
        s.Replace("[|", "[")
         .Replace("|]", "]")
         .Replace("; ", " ")
         .Replace(";", "")
         .Replace("\"", "")
let _floordiv (a:int) (b:int) : int =
    let q = a / b
    let r = a % b
    if r <> 0 && ((a < 0) <> (b < 0)) then q - 1 else q
let mutable _seed: int = 1
let rec set_seed (s: int) =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    let mutable s = s
    try
        _seed <- s
        __ret
    with
        | Return -> __ret
and randint (a: int) (b: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable a = a
    let mutable b = b
    try
        _seed <- int ((((int64 ((_seed * 1103515245) + 12345)) % 2147483648L + 2147483648L) % 2147483648L))
        __ret <- (((_seed % ((b - a) + 1) + ((b - a) + 1)) % ((b - a) + 1))) + a
        raise Return
        __ret
    with
        | Return -> __ret
and jacobi_symbol (random_a: int) (number: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable random_a = random_a
    let mutable number = number
    try
        if (random_a = 0) || (random_a = 1) then
            __ret <- random_a
            raise Return
        random_a <- ((random_a % number + number) % number)
        let mutable t: int = 1
        while random_a <> 0 do
            while (((random_a % 2 + 2) % 2)) = 0 do
                random_a <- _floordiv (int random_a) (int 2)
                let r: int = ((number % 8 + 8) % 8)
                if (r = 3) || (r = 5) then
                    t <- -t
            let temp: int = random_a
            random_a <- number
            number <- temp
            if ((((random_a % 4 + 4) % 4)) = 3) && ((((number % 4 + 4) % 4)) = 3) then
                t <- -t
            random_a <- ((random_a % number + number) % number)
        if number = 1 then
            __ret <- t
            raise Return
        __ret <- 0
        raise Return
        __ret
    with
        | Return -> __ret
and pow_mod (``base``: int) (exp: int) (``mod``: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable ``base`` = ``base``
    let mutable exp = exp
    let mutable ``mod`` = ``mod``
    try
        let mutable result: int = 1
        let mutable b: int = ((``base`` % ``mod`` + ``mod``) % ``mod``)
        let mutable e: int = exp
        while e > 0 do
            if (((e % 2 + 2) % 2)) = 1 then
                result <- (((result * b) % ``mod`` + ``mod``) % ``mod``)
            b <- (((b * b) % ``mod`` + ``mod``) % ``mod``)
            e <- _floordiv (int e) (int 2)
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
and solovay_strassen (number: int) (iterations: int) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable number = number
    let mutable iterations = iterations
    try
        if number <= 1 then
            __ret <- false
            raise Return
        if number <= 3 then
            __ret <- true
            raise Return
        let mutable i: int = 0
        while i < iterations do
            let a: int = randint (2) (number - 2)
            let x: int = jacobi_symbol (a) (number)
            let y: int = pow_mod (a) (_floordiv (int (number - 1)) (int 2)) (number)
            let mutable mod_x: int = ((x % number + number) % number)
            if mod_x < 0 then
                mod_x <- mod_x + number
            if (x = 0) || (y <> mod_x) then
                __ret <- false
                raise Return
            i <- i + 1
        __ret <- true
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        ignore (set_seed (10))
        ignore (printfn "%s" (_str (solovay_strassen (13) (5))))
        ignore (printfn "%s" (_str (solovay_strassen (9) (10))))
        ignore (printfn "%s" (_str (solovay_strassen (17) (15))))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
ignore (main())

File: tests/algorithms/x/PHP/maths/solovay_strassen_primality_test.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _intdiv($a, $b) {
    if ($b === 0 || $b === '0') {
        throw new DivisionByZeroError();
    }
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv(intval($a), intval($b));
}
$__start_mem = memory_get_usage();
$__start = _now();
  $seed = 1;
  function set_seed($s) {
  global $seed;
  $seed = $s;
};
  function randint($a, $b) {
  global $seed;
  $seed = ($seed * 1103515245 + 12345) % 2147483648;
  return ($seed % ($b - $a + 1)) + $a;
};
  function jacobi_symbol($random_a, $number) {
  global $seed;
  if ($random_a == 0 || $random_a == 1) {
  return $random_a;
}
  $random_a = $random_a % $number;
  $t = 1;
  while ($random_a != 0) {
  while ($random_a % 2 == 0) {
  $random_a = _intdiv($random_a, 2);
  $r = $number % 8;
  if ($r == 3 || $r == 5) {
  $t = -$t;
}
};
  $temp = $random_a;
  $random_a = $number;
  $number = $temp;
  if ($random_a % 4 == 3 && $number % 4 == 3) {
  $t = -$t;
}
  $random_a = $random_a % $number;
};
  if ($number == 1) {
  return $t;
}
  return 0;
};
  function pow_mod($base, $exp, $mod) {
  global $seed;
  $result = 1;
  $b = $base % $mod;
  $e = $exp;
  while ($e > 0) {
  if ($e % 2 == 1) {
  $result = ($result * $b) % $mod;
}
  $b = ($b * $b) % $mod;
  $e = _intdiv($e, 2);
};
  return $result;
};
  function solovay_strassen($number, $iterations) {
  global $seed;
  if ($number <= 1) {
  return false;
}
  if ($number <= 3) {
  return true;
}
  $i = 0;
  while ($i < $iterations) {
  $a = randint(2, $number - 2);
  $x = jacobi_symbol($a, $number);
  $y = pow_mod($a, _intdiv(($number - 1), 2), $number);
  $mod_x = $x % $number;
  if ($mod_x < 0) {
  $mod_x = $mod_x + $number;
}
  if ($x == 0 || $y != $mod_x) {
  return false;
}
  $i = $i + 1;
};
  return true;
};
  function main() {
  global $seed;
  set_seed(10);
  echo rtrim(_str(solovay_strassen(13, 5))), PHP_EOL;
  echo rtrim(_str(solovay_strassen(9, 10))), PHP_EOL;
  echo rtrim(_str(solovay_strassen(17, 15))), PHP_EOL;
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
